preprocessing/data_science_help_functions.py

```diff
@@ -189,3 +197,178 @@ def detect_id_target_problem(dataframes_dict: dict, threshold: float = 0.1) -> T
+    :param float threshold: A value larger than 0 and less than 1. If the result of calculation is greater than threshold - there is sugnificant difference between train and test data
+
+    :return:
+            - adversarial_validation_result: Adversarial validation score.
+    """
+
+    # Check if it only one dataframe provided
+    if len(dataframe_dict) != 2:
+        # do nothing and return the original data
+        print("Can't apply adversarial_validation because count of dataframes is not equal to 2")
+        return
+
+    # TODO: support > 2 dataframes
+    # if 2 dataframe than it will be considered as `train` and `test`
+    # TODO: replace to take_first_n from https://docs.python.org/3.8/library/itertools.html#itertools-recipes
+    label_iter = iter(dataframe_dict.keys())
+    train_label = next(label_iter)
+    test_label = next(label_iter)
+
+    train = dataframe_dict[train_label]
+    test = dataframe_dict[test_label]
+
+    df_joined = join_dataframe_for_validation(ignore_columns, max_dataframe_length, test, train)
+
+    # a new target
+    y = df_joined['istrain']
+    df_joined.drop('istrain', axis=1, inplace=True)
+
+    # train classifier
+    adversarial_validation_result, clf = get_adv_validation_score(df_joined, y)
+
+    # result printing
+    def print_template(result, train_label, test_label, adversarial_validation_result, threshold)  -> str:
+        return f"{result} significant difference between {train_label} and {test_label}\n" \
+                      f"datasets in terms of feature distribution. Validation score: {adversarial_validation_result}, threshold: {threshold}"
+
+    # Process result:
+    if adversarial_validation_result > threshold:
+        result = 'There is no'
+        print_template(result, train_label, test_label, adversarial_validation_result, threshold)
+        print(f"Top features are: {xgb_important_features(clf)}")
+    else:
+        result = 'WARNING!!!! There is'
+        print_template(result, train_label, test_label, adversarial_validation_result, threshold)
+    return adversarial_validation_result
+
+
+def join_dataframe_for_validation(ignore_columns: list,
+                                  max_dataframe_length: int,
+                                  test: pd.DataFrame,
+                                  train: pd.DataFrame) -> pd.DataFrame:
+
+    """Join dataframe for validation
+
+    :param list ignore_columns: List of column to ignore (ID, target, etc...)
+    :param int max_dataframe_length: a limit of dataframe length before joining
+    :param pd.DataFrame train,test: Dataframes to join
+
+    :return:
+            - df_joined: Joined dataframe.
+    """
+
+    if len(ignore_columns) > 0:
+        columns_to_use = [x for x in list(test.columns) if x not in ignore_columns]
+        train = train[columns_to_use]
+        test = test[columns_to_use]
+    # add identifier and combine
+    train['istrain'] = 1
```

Comment:
Could you please check this again. do you need to repeat the code here. Let us discuss this if the answer is yes.